smart-contracts/osmosis/contracts/cl-vault/src/vault/swap.rs

```diff
@@ -123,75 +147,132 @@ pub fn execute_swap_non_vault_funds(
+    lower_tick: i64,
+    upper_tick: i64,
+) -> Result<(Coin, SwapDirection, Uint128), ContractError> {
+    if !tokens_provided.0.is_zero() {
+        // range is above current tick
+        let token_in = if pool_details.current_tick > upper_tick {
+            Coin {
+                denom: pool_config.token0.clone(),
+                amount: tokens_provided.0,
+            }
+        } else {
+            Coin {
+                denom: pool_config.token0.clone(),
+                amount: get_single_sided_deposit_0_to_1_swap_amount(
+                    tokens_provided.0,
+                    lower_tick,
+                    pool_details.current_tick,
+                    upper_tick,
+                )?,
+            }
+        };
+        let left_over_amount = tokens_provided.0.checked_sub(token_in.amount)?;
+        Ok((token_in, SwapDirection::ZeroToOne, left_over_amount))
+    } else {
+        // current tick is above range
+        let token_in = if pool_details.current_tick < lower_tick {
+            Coin {
+                denom: pool_config.token1.clone(),
+                amount: tokens_provided.1,
+            }
+        } else {
+            Coin {
+                denom: pool_config.token1.clone(),
+                amount: get_single_sided_deposit_1_to_0_swap_amount(
+                    tokens_provided.1,
+                    lower_tick,
+                    pool_details.current_tick,
+                    upper_tick,
+                )?,
+            }
+        };
+        let left_over_amount = tokens_provided.1.checked_sub(token_in.amount)?;
+        Ok((token_in, SwapDirection::OneToZero, left_over_amount))
+    }
+}
+
 #[allow(clippy::too_many_arguments)]
 pub fn calculate_swap_amount(
-    deps: DepsMut,
+    deps: &DepsMut,
     env: &Env,
-    pool_config: PoolConfig,
     swap_direction: SwapDirection,
-    token_in_amount: Uint128,
+    token_in: Coin, // this is a coin so we can pass external funds as token_in
     max_slippage: Decimal,
+    swap_pool_id: Option<u64>,
     forced_swap_route: Option<Vec<SwapAmountInRoute>>,
-    twap_window_seconds: u64,
+    twap_price: Decimal,
 ) -> Result<SwapCalculationResult, ContractError> {
-    let twap_price = get_twap_price(deps.storage, &deps.querier, env, twap_window_seconds)?;
-    let (token_in_denom, token_out_denom, token_out_ideal_amount) = match swap_direction {
-        SwapDirection::ZeroToOne => (
-            &pool_config.token0,
+    let pool_config = POOL_CONFIG.load(deps.storage)?;
+
+    // Determine the target token and amount based on swap direction
+    let (denom_out, amount_out_ratio) = match swap_direction {
+        SwapDirection::ZeroToOne | SwapDirection::AnyToOne => (
             &pool_config.token1,
-            token_in_amount
-                .checked_multiply_ratio(twap_price.numerator(), twap_price.denominator()),
+            (twap_price.numerator(), twap_price.denominator()), // TODO: Check if this is correct order for AnyToOne
```

Comment:
As in the previous two comments, we should check that the way we use twap_price here is consistent with how we inject arguments from above. The exact_deposit and any_deposit workflows are hardcoding the right assets on the requests, as we know we always work with asset0 and asset1 (base and quote). This is not the case for the SwapNonVaultFunds, which are accepting arbitrary routes and token ins from outside, and this could be hijackable.